lib/elixir/src/elixir_erl_pass.erl

```diff
@@ -361,21 +362,30 @@ translate_map(Meta, Assocs, S) ->
 translate_struct(Meta, Name, {'%{}', _, [{'|', _, [Update, Assocs]}]}, S) ->
   Ann = ?ann(Meta),
   Generated = erl_anno:set_generated(true, Ann),
-  {VarName, _, VS} = elixir_erl_var:build('_', S),
-
-  Var = {var, Ann, VarName},
-  Map = {map, Ann, [{map_field_exact, Ann, {atom, Ann, '__struct__'}, {atom, Ann, Name}}]},
-
-  Match = {match, Ann, Var, Map},
-  Error = {tuple, Ann, [{atom, Ann, badstruct}, {atom, Ann, Name}, Var]},
+  {TUpdate, US} = translate_arg(Update, S, S),
 
-  {TUpdate, US} = translate_arg(Update, VS, VS),
-  {TAssocs, TS} = translate_map(Meta, Assocs, {ok, Var}, US),
+  case elixir_erl:get_type(TUpdate, US) of
+    {struct, Name} ->
+      translate_map(Meta, Assocs, {ok, TUpdate}, US);
+    MaybeMatches when MaybeMatches =:= map orelse MaybeMatches =:= term ->
+      {VarName, _, VS} = elixir_erl_var:build('_', US),
+
+      Var = {var, Ann, VarName},
+      Map = {map, Ann, [{map_field_exact, Ann, {atom, Ann, '__struct__'}, {atom, Ann, Name}}]},
+
+      Match = {match, Ann, Var, Map},
+      Error = {tuple, Ann, [{atom, Ann, badstruct}, {atom, Ann, Name}, Var]},
+
+      {TAssocs, TS} = translate_map(Meta, Assocs, {ok, Var}, VS),
+      {{'case', Generated, TUpdate, [
+        {clause, Ann, [Match], [], [TAssocs]},
+        {clause, Generated, [Var], [], [elixir_erl:remote(Ann, erlang, error, [Error])]}
+      ]}, TS};
+    _WontMatch ->
```

Comment:
Btw, you are probably aware the code above is deprecated. And unfortunately it is the source of a lot of complexity. I have considered releasing Elixir v2.0 multiple times exactly because of that.